autoload/maktaba/keymapping.vim

```diff
@@ -0,0 +1,284 @@
+""
+" @dict KeyMapping
+" A maktaba representation of a vim key mapping, which is used to configure and
+" unmap it from vim.
+
+
+if !exists('s:next_keymap_id')
+  let s:next_keymap_id = 0
+endif
+
+if !exists('s:KEYMAPS_BY_ID')
+  let s:KEYMAPS_BY_ID = {}
+endif
+
+
+function! s:ReserveKeyMapId() abort
+  let l:keymap_id = s:next_keymap_id
+  let s:next_keymap_id += 1
+  return l:keymap_id
+endfunction
+
+
+function! s:GetMappingById(id) abort
+  try
+    return s:KEYMAPS_BY_ID[a:id]
+  catch /E716:/
+    return 0
+  endtry
+endfunction
+
+
+function! s:GetFuncCallKeystrokes(funcstr, mode) abort
+  if a:mode ==# 'n'
+    return printf(':call %s<CR>', a:funcstr)
+  elseif a:mode ==# 'i'
+    return printf('<C-\><C-o>:call %s<CR>', a:funcstr)
+  elseif a:mode ==# 'v'
+    " Uses "gv" at the end to re-enter visual mode.
+    return printf(':<C-u>call %s<CR>gv', a:funcstr)
+  elseif a:mode ==# 's'
+    " Uses "gv<C-g>" at the end to re-enter select mode.
+    return printf('<C-\><C-o>:<C-u>call %s<CR>gv<C-g>', a:funcstr)
+  endif
+  throw maktaba#error#NotImplemented(
+      \ 'MapOnce not implemented for mode %s', a:mode)
+endfunction
+
+
+""
+" @dict KeyMapping
+" Unmaps the mapping in vim.
+" Returns 1 if mapping was found and unmapped, 0 if mapping was gone already.
+function! maktaba#keymapping#Unmap() dict abort
+  if self.IsMapped()
+    let l:arg_prefix = self._maparg.buffer ? '<buffer> ' : ''
+    execute printf(
+        \ 'silent %sunmap %s%s',
+        \ self._maparg.mode,
+        \ l:arg_prefix,
+        \ self._maparg.lhs)
+    if has_key(s:KEYMAPS_BY_ID, self._id)
+      unlet s:KEYMAPS_BY_ID[self._id]
+    endif
+    return 1
+  else
+    return 0
+  endif
+endfunction
+
+
+""
+" @dict KeyMapping
+" Returns 1 if the mapping is still defined, 0 otherwise
+"
+" Caveat: This detection can currently false positive if the original mapping
+" was unmapped but then another similar one mapped afterwards.
```

Comment:
It'd have to be identical, wouldn't it? (same RHS)

Reply:
Yeah, sounds right. I removed this caveat.